File: typescript_version/src/services/messageService.ts
import crypto from 'crypto';
import { Message } from '../types';

// In-memory storage for messages (in production, use a proper database)
const messages = new Map<string, Message[]>();

class MessageService {
    constructor() {
        // Clean up expired messages every minute
        setInterval(this.cleanupExpiredMessages.bind(this), 60000);
    }

    // Store a message for a user
    storeMessage(userId: string, message: string, encryptedMessage: string): Message {
        if (!messages.has(userId)) {
            messages.set(userId, []);
        }

        const messageData: Message = {
            id: crypto.randomBytes(16).toString('hex'),
            message: encryptedMessage,
            timestamp: new Date().toISOString(),
            expiresAt: new Date(Date.now() + 10 * 60 * 1000) // 10 minutes expiry
        };

        messages.get(userId)!.push(messageData);
        return messageData;
    }

    // Get all messages for a user
    getMessages(userId: string): Message[] {
        const userMessages = messages.get(userId) || [];
        return userMessages.filter(msg => !this.isExpired(msg));
    }

    // Check if a message is expired
    private isExpired(message: Message): boolean {
        return new Date(message.expiresAt) < new Date();
    }

    // Clean up expired messages
    private cleanupExpiredMessages(): void {
        for (const [userId, userMessages] of messages.entries()) {
            messages.set(userId, userMessages.filter(msg => !this.isExpired(msg)));
        }
    }
}

export default new MessageService(); 

File: python_version/services/message_service.py
import os
import time
from typing import Dict, List
from dataclasses import dataclass
from datetime import datetime, timedelta

@dataclass
class Message:
    id: str
    message: str
    timestamp: str
    expires_at: datetime

class MessageService:
    def __init__(self):
        self.messages: Dict[str, List[Message]] = {}
        self._cleanup_expired_messages()

    def store_message(self, user_id: str, message: str, encrypted_message: str) -> Message:
        """Store a message for a user."""
        if user_id not in self.messages:
            self.messages[user_id] = []

        message_data = Message(
            id=os.urandom(16).hex(),
            message=encrypted_message,
            timestamp=datetime.utcnow().isoformat(),
            expires_at=datetime.utcnow() + timedelta(minutes=10)
        )

        self.messages[user_id].append(message_data)
        return message_data

    def get_messages(self, user_id: str) -> List[Message]:
        """Get all non-expired messages for a user."""
        self._cleanup_expired_messages()
        return self.messages.get(user_id, [])

    def _is_expired(self, message: Message) -> bool:
        """Check if a message is expired."""
        return datetime.utcnow() > message.expires_at

    def _cleanup_expired_messages(self) -> None:
        """Clean up expired messages."""
        for user_id in list(self.messages.keys()):
            self.messages[user_id] = [
                msg for msg in self.messages[user_id]
                if not self._is_expired(msg)
            ]
            if not self.messages[user_id]:
                del self.messages[user_id]

message_service = MessageService() 